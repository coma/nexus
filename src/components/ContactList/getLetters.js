const letters = [...'ABCDEFGHIJKLMNOPQRSTUVWXYZ'];
const other = /^[^a-z]/i;

export default function getLetters(contacts) {
  const findId = predicate => {
    const match = contacts.find(predicate);
    return match && match.id;
  };

  const indexes = letters.map(letter => ({
    letter,
    id: findId(contact => contact.name.first.toUpperCase().startsWith(letter)),
  }));

  indexes.push({
    letter: '#',
    id: findId(contact => other.test(contact.name.first)),
  });

  return indexes;
}
