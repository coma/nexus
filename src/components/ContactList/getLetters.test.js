import getLetters from './getLetters';

describe('The getLetters helper', () => {
  it('should index by letters and others', () => {
    expect(
      getLetters([
        { id: '3', name: { first: '3' } },
        { id: 'a', name: { first: 'a' } },
        { id: 'x', name: { first: 'X' } },
      ]),
    ).toMatchSnapshot();
  });
});
