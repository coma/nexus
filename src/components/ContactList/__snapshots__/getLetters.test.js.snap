// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The getLetters helper should index by letters and others 1`] = `
Array [
  Object {
    "id": "a",
    "letter": "A",
  },
  Object {
    "id": undefined,
    "letter": "B",
  },
  Object {
    "id": undefined,
    "letter": "C",
  },
  Object {
    "id": undefined,
    "letter": "D",
  },
  Object {
    "id": undefined,
    "letter": "E",
  },
  Object {
    "id": undefined,
    "letter": "F",
  },
  Object {
    "id": undefined,
    "letter": "G",
  },
  Object {
    "id": undefined,
    "letter": "H",
  },
  Object {
    "id": undefined,
    "letter": "I",
  },
  Object {
    "id": undefined,
    "letter": "J",
  },
  Object {
    "id": undefined,
    "letter": "K",
  },
  Object {
    "id": undefined,
    "letter": "L",
  },
  Object {
    "id": undefined,
    "letter": "M",
  },
  Object {
    "id": undefined,
    "letter": "N",
  },
  Object {
    "id": undefined,
    "letter": "O",
  },
  Object {
    "id": undefined,
    "letter": "P",
  },
  Object {
    "id": undefined,
    "letter": "Q",
  },
  Object {
    "id": undefined,
    "letter": "R",
  },
  Object {
    "id": undefined,
    "letter": "S",
  },
  Object {
    "id": undefined,
    "letter": "T",
  },
  Object {
    "id": undefined,
    "letter": "U",
  },
  Object {
    "id": undefined,
    "letter": "V",
  },
  Object {
    "id": undefined,
    "letter": "W",
  },
  Object {
    "id": "x",
    "letter": "X",
  },
  Object {
    "id": undefined,
    "letter": "Y",
  },
  Object {
    "id": undefined,
    "letter": "Z",
  },
  Object {
    "id": "3",
    "letter": "#",
  },
]
`;
